features/0434-outofband/README.md

```diff
@@ -0,0 +1,417 @@
+
+The *receiver* state machine is as follows:
+
+[![receiver state machine google doc](state-machine-receiver.png)](https://docs.google.com/spreadsheets/d/1o1szGCMx6FzrxfAKV76IEYmeKRJqvV3gd1r3Qfv1Q9o/edit#gid=10874521)
+
+Worth noting is the first event of the `done` state, where the receiver may receive the message multiple times. This represents, for example, an agent returning to the same website and being greeted with the same QR code each time.
+
+### Messages
+
+There are two messages in the out of band protocol. They are both sent by the *sender*. Only one of the message types is ever used in an instance of the protocol.
+
+#### Message Type: `https://didcomm.org/outofband/VER/invitation`
+
+```jsonc
+{
+  "@type": "https://didcomm.org/outofband/VER/invitation",
+  "@id": "<id used for context as pthid>",
+  "label": "Faber College",
+  "protocols": [
+      "https://didcomm.org/didexchange/1.0",
+      "https://didcomm.org/connections/1.0"
+      ]
+  "service": ["did:sov:LjgpST2rjsoxYegQDRm7EL"]
+}
+```
+
+The items in the message are:
+
+- `@type` - the DIDComm message type
+- `@id` - the unique ID of the message. The ID should be used as the **parent** thread ID (`pthid`) for the response message, rather than the more common thread ID (`thid`) of the response message. This enables multiple uses of a single `invitation` message.
+- `label` - a self-attested string that the receiver may want to display to the user.
+- `protocols` - array of protocols in the order of preference of the sender that the receiver can use in responding to the message. As implied by the message type of invitation, these are not arbitrary protocols but rather protocols that result in the establishment of a connection.
+- `service` - an item that is the equivalent of the service block of a DIDDoc that the receiver is to use in responding to the message. Additional details below.
+
+While the _receiver_ is expected to respond with an initiating message from the chosen protocol using a specified service, the receiver may respond by reusing an existing connection. Specifically, if a connection they have was created from an out of band `invitation` from the same public DID of a new `invitation`, the receiver **SHOULD** use the existing connection in responding to the `invitation`. The selected `protocol` being used by the receiver may have a specific message type to use for this purpose.
+
+> **To Do**: Can we expand this to be able to reuse an inline `service` entry?
+> **To Do**: Update the `did-exchange` (and perhaps `connection`) protocol to have a `continue` message to be used for this purpose.
+
+#### Message Type: `https://didcomm.org/outofband/VER/request`
+
+```jsonc
+{
+  "@type": "https://didcomm.org/outofband/VER/request",
```

Comment:
After reading the doc, I'm puzzled about the difference between these two message types. Do we really need both of them. Aren't they doing almost the same thing, except that the request message attaches some stuff? Is that difference worth a whole separate message?

A possible answer to my question might be: no, they are different. The invitation just asks to connect, whereas the request message can do arbitrary stuff. But if that's true, why shouldn't we use the arbitrary stuff to do connections? I can already attach things to an invitation anyway, since it's a didcomm message. Right?

Reply:
I don't think it is cut'n'dry that two are needed, but my opinion is that the two are different enough that it's worth having both.

Very little information is needed is needed in the `invitation` message because it's purpose is known; quite a lot is needed in the first `request` example we have - `presentation-request`, especially if we are going to support multiple credential exchange models. I expect other uses to likely be equally bulky.  I can't think of what the attachment would be contained in the `invitation` case, so it just wouldn't have it.  I think we want to make it very obvious to the controller dev if a connection is to be established or not as an outcome, and having the two messages accomplish that.

I think the benefits are there for the two.  Others?

Reply:
Agree on two. I think the benefit of generalization is lost for common cases (such as establishing a connection) where a dedicated message is both smaller and harder to screw up.

Reply:
>Very little information is needed is needed in the invitation message because it's purpose is known

This is the disconnect. I'm claiming that the *real* purpose of the invitation is not known, and that the purpose we're currently assuming (just to connect for the sake of connecting) is indicative of immaturity of our thinking as a community; random agents who want to connect just for the sake of connecting should always be refused when the ecosystem is mature. We only think that's valid today because it's what we've been using to build a narrative for our demos. The goal of just connecting is almost never the real purpose. And I'm asking to make the real purpose known (which happens to eliminate the distinction with a `request` message), and getting lots of pushback.

We can certainly merge this protocol RFC, but I can't recommend to my own company or anyone in the community that we support it. It's DOA for us unless we take the spam issue more seriously.

Reply:
As per above, the invitee knows the purpose based on the action of ingesting an out of band message.  Totally agree that the purpose of the invitation is not to connect, but to do something after, but adding more self-attested data such as the name of a protocol to use next does not provide more context in my opinion. 

Daniel, we *really* don't want this to be DOA/vetoed, so how do we get around this.  Could you:

* explain in very concrete term spam examples -- how you see them happening.
* exactly what you want added to the invitation message?

On your latter point ("which happens to eliminate the distinction..."), my guess is what you will propose will not eliminate the request message, but I'll wait on that until I see exactly what it is you are proposing.

Maybe we can discuss at HGF?  Are you still here on Friday?

Reply:
@swcurran , @andrewwhitehead , @llorllale , and @TelegramSam : here's my best attempt to provide more context about my spam concern: https://docs.google.com/document/d/1S3dC9NuIYx9IRJVWu5RUG1wLNluhVy0lhIinin6fRqw/edit

Reply:
Daniel, thanks for that writeup. I'm thinking that clarifying that outofband messages are for human reception and contextual evaluation solves the recipient spam problem sufficient for current needs. Non-human reception can be expanded in a future version that includes greater features to prevent the issues you highlight, or in a different protocol alltogether.

Future version of the protocol can also add additional features necessary for contextual evaluation if we deem the current approach insufficient.

Reply:
That doesn't resolve my concern. Two reasons.

1. I'm thinking this "human only" guidance is not aligned with the intentions of @tplooker WRT ephemeral use cases. Isn't he planning on using ephemeral stuff with IoT devices, immediately?

2. I don't think we get to say that protocols are only for use by people. DIDComm is designed such that agents could be backed by people or by AIs or by automation, and we can't tell the difference; everything is a peer. If there is some technical reason why the protocol *can't* be used by tech, as opposed to simple intentions that it not be, that would make me feel more cheerful.

Reply:
And reason 3: the human flooding problem isn't solved.

Reply:
@TelegramSam : what would resolve my concern is helping me realize that one of my assumptions is wrong. That could totally be the case. If it is, great. But if not, then a general statement that we can tamp down the worry about spam by telling senders and recipients to be wise isn't going to help me get comfortable. I have an immediate use case for this protocol in a context where millions of people will receive unsolicited invitations. These invitations will be associated with the reputation of an institution that cares enormously about its public trust. Furthermore, once this protocol is deployed in this use case, it will be difficult to upgrade it. So the "future version can add additional features if we deem the current approach insufficient" is already a fulfilled prophecy for me. I need a better solution, now.

I also truly don't get the complexity concern. What I proposed adds one requirement: the existence of a field in a message called `goals`. The field can have no content. The recipient of such a message can ignore the content if they like; nothing in the protocol requires either party to do anything significant with the field. Anybody who doesn't want the complexity can largely ignore it in their impls. So how does this add complexity is such an obnoxious way?

Reply:
> So how does this add complexity is such an obnoxious way?

Goals don't add complexity, generalizing/collapsing the two message types does. 

I do find goals to be fairly useless in spam prevention, given that they can just be lies.

> And reason 3: the human flooding problem isn't solved.

I don't find this problem to be real. In the AR example (from the linked doc), I find it silly to think that Alice's AR platform would not be capable of recording when and where (the context) that an outofband message was received.

> I don't think we get to say that protocols are only for use by people.

I think we can say that _this_ protocol is only for use by people. This prevents developers from needing to receive these messages over any inbound transport.

> I have an immediate use case for this protocol in a context where millions of people will receive unsolicited invitations. These invitations will be associated with the reputation of an institution that cares enormously about its public trust.

Knowing more about this use case will help us evaluate the design of the protocol. How are the invitations received? How do goals help establish trust?

Reply:
> I'm thinking this "human only" guidance is not aligned with the intentions of @tplooker WRT ephemeral use cases. Isn't he planning on using ephemeral stuff with IoT devices, immediately? (Tobias, please comment. Maybe this concern is not really a thing?)

Also interested in commentary from @tplooker to verify this.

Reply:
Please read this comment carefully, and please respond carefully. I feel like I am being misunderstood repeatedly, and that a quick response will only make things worse. I will number the items in this comment so they are easier to refer to with shorthand.

1. I am not recommending that we collapse the two message types. If you go read my comments, you will see that I'm puzzled about the `request` message. I have also pointed out in a couple verbal contexts that it's hard to justify such a message if the `invitation` message becomes richer. And I've said in this PR that the purpose of it is pretty hard to understand from the written text. But I haven't advocated anywhere in this PR stream that we get rid of `request`. I have made one concrete proposal only: that we add `goal` as a property to the `invitation` message. And the pushback I'm getting is about complexity. @TelegramSam says he's not troubled by the complexity of having a `goal` field--only about collapsing the two message types, but I think @swcurran feels that `goal` is too complex even if we keep the message types separate. Is that true? If so, why? If not, can we please accept that suggestion? I don't feel nearly as interested in collapsing message types as in getting a goal into an invitation.

2. Sam says "I think we can say that this protocol is only for use by people. This prevents developers from needing to receive these messages over any inbound transport." But the key property of agents is that they represent people in a way that parties on the other side of the interaction don't get to dictate. We don't get to require Alice to process the code with her human intelligence; she can use an AI or a dumb policy or whatever she wants.

    Also, we don't ever get to assume in DIDComm that the transport used by the sender and the transport used by the receiver are identical. As a sender, you can certainly choose the QR code transport -- but you don't get to tell Alice that she can't convert a QR code to an emailed link, or a voice text, or whatever, via a relay. (My wife has a deaf friend. She calls on the phone; a relay converts her spoken words to typed text. How will a blind person process a QR code? Etc.) All you know as a sender is that you *sent* over the QR code channel. Alice will receive it however she wants, and will process it however she wants. This protocol would (and should) work over any transport. I gave several transport examples in my doc that explored the spam problem; I claim all of them are reasonable. It is possible that a protocol could be invented that is inherently transport-specific (e.g., a protocol that lets jazz musicians jam together requiring realtime audio steaming and disallowing pigeon transport), but the "let's connect" problem is not that sort of problem, and we shouldn't do headstands to bend it into one. Agents don't have to support transports they don't like, but protocols aren't written for the convenience of a few agents that want to be less than transport agnostic.

    Also, even if you throw all of that out the window, what property of the transport are you going to modify to declare it to be a human-only transport? Just verbiage that a human reads in an RFC? If so, what happens when you want to change your mind later? Does that require a version update? If so, why, since none of the machine-inspectable properties of the protocol are different? And if not, how will you tell whether you're dealing with an agent that expects a human partner or not?

    Bottom line: saying this protocol is for humans only is a BAD idea for lots of reasons.

3. Sam says, "I do find goals to be fairly useless in spam prevention, given that they can just be lies."

    A lie is just as good as a truth here, because *what we're after is something that becomes testable quickly, not something that's guaranteed to be true*. Suppose Alice gets invitation A with no goal, invitation B with the goal of issuing a credential, and invitation C with the goal of giving her free money. Suppose B is a truthful invitation, and C is a lie. Alice responds with a connection request to each one, and gets a connection response, so she's now built a connection for all 3 invitations. What happens next? The sender of A (no goal) is silent. Alice waits. How long should she wait? Forever -- since the sender isn't violating her expectations. The sender of B (truthful invitation to issue cred) immediately starts an issuance protocol. Alice is happy about B and knows she isn't wasting her time. How about the liar who sent invitation C? If the liar is silent, Alice deletes the connection after, say, 30 seconds, since they didn't behave according to their expressed goal. If the liar now tries to do anything other than start the free money protocol (e.g., sell Alice Viagra, but also ask Alice to prove something, propose a rendezvous, send a BasicMessage, etc) Alice can also immediately delete the connection, since she now knows C is a liar. The bad outcome here isn't C (that's just a temporary annoyance, whether the party lies or not); rather, the bad outcome is A (no expressed goal). Because now Alice has allocated a DID, keys, a peer DID doc, etc. Having one of these useless connections is no big deal, but having thousands of them might become problematic.

    There will always be connections that have no declared purpose. But adding a `goal` field lets Alice filter unsolicited invitations twice. First she throws away ones with goals she either dislikes or can't accomplish. Then she quickly and effortlessly throws away lies about intention. This can be done in automated fashion, without bothering her at all, if there's a goal. If not, Alice is forced to triage them all personally.

4. Sam says, "I don't find this problem to be real. In the AR example (from the linked doc), I find it silly to think that Alice's AR platform would not be capable of recording when and where (the context) that an outofband message was received."

I feel like this kind of misses the point. Of course her AI platform can record when and where an OOB message was received. But if it's just an invitation to connect, without any goal, then there's no data for her AI to reason from, other than when and where the invitation occurred. Was that jewelry store going to offer her a 90% discount on a second set of earrings if she's a returning customer -- redeemable any time she comes by this week? She might want to know about that when she gets home. Whereas she may have no interest in the Hog Heaven Road Gear shop she walked past, since that was touting a giveway of a free Harley but it was a time-limited offer that has expired. The point is that the context includes the time and circumstances when the invitation arrived, but that gets way more useful when software knows the asserted intention of the invitation. Now policy can operate on intention.

Reply:
My $0.02CDN.  I added some comments to the Google Doc that shows where I don't agree with either assumptions or the scenarios. The biggest disagreement I have with the doc is that assumption that there will be an automatic response to an OOB message. Human or machine, a (possibly automated) decision will be made between receiving and accepting the OOB message. The bottom line is that I still don't see spam as a compelling concern with respect to this protocol.

If we want to add anything to address this, I think it would be preferred to add a "purpose" text/code field (with, of course, the option of ~l10n). Like Sam, I don't think it is needed and that it is potentially misleading because it is self-asserted.  The use of "goals" list of protocols Daniel proposes gives the human no help in knowing what the initial purpose of the connection is. For a machine it does (but no different than a code), but the machine handling will be conceptually different from that of a human.

The reason I think a text field might be useful is that the "label" field will become overloaded for that purpose. I expect the label will be used as an initial value for the connection (update-able by the user), and having the initial purpose for the connection in the label will be a pain. So separating them into two fields could be useful.